Issue676

YizhangCao opened the pull request:
# Homework 2 Pull Request

## Checklist before creating

- [ x ] There are no Checkstyle issues.
- [ x ] There are no SpotBugs issues.
- [ x ] There are no javadoc errors.
- [ x ] All tests in the file pass.

_Please explain if there are any issues/errors._

## Questions

### Will you get full credit on this assignment?

Yes, I converted all assigned multi-assertion test methods in AuthorListTest and AuthorTest to use @ParameterizedTest, followed JabRef conventions, and addressed Javadoc/Checkstyle issues.

### How long did you spend on the assignment?
6-7 hours
### What problems did you encounter, and how did you deal with them?
Problem: Javadoc @link caused doclint errors.
Solution: Changed it to {@code …}.
Bug:CSV/parameterization errors: “Found: boolean, required: java.lang.String”
Fix:make each row is one CSV record and that all strings are quoted
### What help did you give or receive on this assignment (including answering these question)? What was your AI usage?
Search on google to understand parameterized tests, and use chat-gpt to check whether meet all rubric.

YizhangCao committed:
Replace method addDotIfAbbreviationAddDot() with a @ParameterizedTest using @CsvSource, so each input/expected pair runs as a separate case.

YizhangCao committed:
Replace method addDotIfAbbreviationAddDot() with a @ParameterizedTest using @CsvSource, so each input/expected pair runs as a separate case.

YizhangCao committed:
Added import org.junit.jupiter.params.ParameterizedTest; and import org.junit.jupiter.params.provider.CsvSource

YizhangCao committed:
replace fixAuthorNatbib parameterized with 4 CSV cases

YizhangCao committed:
replace fixAuthorFirstNameFirstCommas →to fixAuthorFirstNameFirstCommas_param(String input, String expected, boolean abbreviate, boolean oxford)

YizhangCao committed:
replace fixAuthorFirstNameFirst to fixAuthorFirstNameFirst_param(String input, String expected)

YizhangCao committed:
replace fixAuthorLastNameFirstCommasOxfordComma to fixAuthorLastNameFirstCommas_oxford(String input, String expected, boolean abbreviate)

YizhangCao committed:
replace fixAuthorLastNameOnlyCommas to fixAuthorLastNameOnlyCommas_param(String input, String expected, boolean oxford)

YizhangCao committed:
change @link to @code for debug

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -16,7 +18,7 @@
 
 /**
- * Other parsing tests are available in {@link org.jabref.logic.importer.AuthorListParserTest}
+ * Other parsing tests are available in {@code org.jabref.logic.importer.AuthorListParserTest}
```
It's great that you fixed javadoc errors that you didn't cause! Please see if you can get the link working.

YizhangCao replied:
Thank you for pointing this out! I originally switched the @link to @code to fix the doclint error, but I see now that the intention was to make the reference clickable.

espertusnu replied:
Can you try fixing that?

YizhangCao replied:
I still back to use link but also linked it with github address

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -46,15 +48,16 @@ public static int size(String bibtex) {
+  @CsvSource({
+      "'', ''",
+      "'John Smith', 'Smith'",
+      "'John Smith and Black Brown, Peter', 'Smith and Black Brown'",
```
Good use of single-quote.

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -119,69 +122,20 @@ void parseCachesOneLatexFreeAuthor() {
-    }
+  @ParameterizedTest
+  @CsvSource({
+      "'John Smith', 'John Smith', false, false",
```
Good. Could you also copy over the comments?

YizhangCao replied:
Done

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -288,90 +242,37 @@ void getAsFirstLastNamesLatexFreeOneInstitutionWithParanthesisAtStart() {
+      "'John Smith and Black Brown, Peter', 'Smith, J. and Black Brown, P.', true",
+      "'John Peter von Neumann', 'von Neumann, J. P.', true"
+  })
+  void fixAuthorLastNameFirstCommas_noOxford(String input, String expected, boolean abbreviate) {
```
I like your test method names, but we should follow the convention in the file we are modifying.

YizhangCao replied:
Understood, have renamed methods

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -288,90 +242,37 @@ void getAsFirstLastNamesLatexFreeOneInstitutionWithParanthesisAtStart() {
+      "John Smith", "John Smith"
+  })
+  void fixAuthorFirstNameFirst_param(String input, String expected) {
+    assertEquals(expected, AuthorList.fixAuthorFirstNameFirst(input));
```
I don't understand why this is in its own method.

YizhangCao replied:
Merge fixAuthorFirstNameFirst_param into main test

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -615,35 +516,21 @@ void fixAuthorLastNameFirst() {
+      "'John Smith and Black Brown, Peter', 'Smith, J. and Black Brown, P.', true",
+      "'John Peter von Neumann', 'von Neumann, J. P.', true"
+  })
+  void fixAuthorLastNameOnlyCommas_param(String input, String expected, boolean oxford) {
```
Drop `_param` here and elsewhere. Reuse the name of the method you are replacing.

YizhangCao replied:
Rename tests to match AuthorListTest convention

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -615,35 +516,21 @@ void fixAuthorLastNameFirst() {
-    }
-
-    @Test
+  @SuppressWarnings("checkstyle:EmptyLineSeparator")
```
Suppressing a warning is not the best way of dealing with it. Have you checked if you can fix the underlying problem?

YizhangCao replied:
Good catch — I removed the suppression and fixed the underlying formatting to satisfy EmptyLineSeparator: there’s now exactly one blank line between methods, and annotations are placed directly above the method without extra blank lines. I also removed an unintended @Test on a parameterized test.

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorTest.java:
```diff
@@ -4,20 +4,24 @@
+        "'AO', 'A. O.'",
+        "'AO.', 'A. O.'",
+        "'A.O.', 'A. O.'",
+        "'A-O', 'A.-O.'"
```
Are the single quotes needed in this case?

YizhangCao replied:
Remove unnecessary quotes in AuthorTest CsvSource

YizhangCao committed:
Restore comments in parameterized tests

YizhangCao committed:
Rename tests to follow AuthorListTest convention

YizhangCao committed:
Merge fixAuthorFirstNameFirst_param into main test

YizhangCao committed:
Merge fixAuthorFirstNameFirst_param into main test

YizhangCao committed:
Merge fixAuthorFirstNameFirst_param into main test

YizhangCao committed:
Merge fixAuthorFirstNameFirst_param into main test

YizhangCao committed:
Merge fixAuthorFirstNameFirst_param into main test

YizhangCao committed:
Merge fixAuthorFirstNameFirst_param into main test

YizhangCao committed:
Merge fixAuthorFirstNameFirst_param into main test

YizhangCao committed:
Rename tests to match AuthorListTest convention

YizhangCao committed:
Rename tests to match AuthorListTest convention

YizhangCao committed:
Good catch — I removed the suppression and fixed the underlying formatting to satisfy EmptyLineSeparator: there’s now exactly one blank line between methods, and annotations are placed directly above the method without extra blank lines. I also removed an unintended @Test on a parameterized test.

YizhangCao committed:
Good catch — I removed the suppression and fixed the underlying formatting to satisfy EmptyLineSeparator: there’s now exactly one blank line between methods, and annotations are placed directly above the method without extra blank lines. I also removed an unintended @Test on a parameterized test.

YizhangCao committed:
Remove unnecessary quotes in AuthorTest CsvSource

espertusnu commented:
Great responses! I'm going to highlight these in class (without your name, unless you give me the okay to use it).

I don't understand why there are so many commit messages. Some have great commit messages, like "Remove unnecessary quotes in AuthorTest CsvSource". Others have messages that don't fit the guidelines (max 50 characters), such as "Good catch..."

Remember to request a review through github when you're ready.

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -124,16 +124,25 @@ void parseCachesOneLatexFreeAuthor() {
+      // One author
       "'John Smith', 'John Smith', false, false",
       "'John Smith', 'J. Smith', true, false",
       "'John Smith and Black Brown, Peter', 'J. Smith and P. Black Brown' true, false",
```
This has two authors

YizhangCao replied:
I moved that case into the “Two authors” group and fixed the missing comma so the CSV parses correctly. The parameterized rows are now grouped and labeled accurately.

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorTest.java:
```diff
@@ -12,14 +12,13 @@
-    })
+  @ParameterizedTest
+  @CsvSource({
+      "AO, 'A. O.'",
```
What's the purpose of the single quotes?

YizhangCao replied:
I’ve removed the unnecessary single quotes around the input values. The quotes remain around the expected values since they contain spaces, which JUnit requires.

YizhangCao committed:
Fix CSV row and grouping in FirstNameFirstCommas

YizhangCao committed:
Fix CSV row and grouping in FirstNameFirstCommas

YizhangCao committed:
Fix Javadoc link to external test file

espertusnu reviewed (changes requested):
I saw that I had a pending review for you. These comments might all be outdated.

espertusnu commented:
I've lost track of whose turn it is. Could you please let me know?

YizhangCao commented:
Sorry about that. I fix @link and @code problem by adding<a href=...> → generates a proper external link. Right now I still use @link with no Cannot resolve symbol error.

YizhangCao committed:
Add <a href=...> → generates a proper external link. No Cannot resolve symbol error right now.

YizhangCao commented:
I have fixed the problem before, But forget to commit it.

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -16,7 +18,9 @@
 /**
- * Other parsing tests are available in {@link org.jabref.logic.importer.AuthorListParserTest}
+ * Other parsing tests are available in
+ * {@link <a href="https://github.com/JabRef/jabref/blob/main/jabref-logic/src/test/java/org/jabref/logic/importer/AuthorListParserTest.java">AuthorListParserTest</a>}.
```
I believe this is incorrect. See https://stackoverflow.com/questions/1082050/linking-to-an-external-url-in-javadoc

YizhangCao replied:
I change back to {@link org.jabref.logic.importer.AuthorListParserTest} and run javadoc without error, and also all tests passed. I think it may due to gradle set up problem or something else

espertusnu commented on jablib/src/test/java/org/jabref/model/entry/AuthorListTest.java:
```diff
@@ -16,7 +18,9 @@
- * Other parsing tests are available in {@link org.jabref.logic.importer.AuthorListParserTest}
+ * Other parsing tests are available in
+ * {@link <a href="https://github.com/JabRef/jabref/blob/main/jabref-logic/src/test/java/org/jabref/logic/importer/AuthorListParserTest.java">AuthorListParserTest</a>}.
+ * See also {@link org.jabref.logic.importer.AuthorListParser AuthorListParser}.
```
Does this link work?

YizhangCao replied:
Yes, this link works correctly without bugs

espertusnu reviewed (changes requested):
We've been going back and forth a lot about the @link. Please schedule a meeting with me or attend my office hours. I think that would be more productive.

espertusnu commented:
I downloaded your code, but `./gradlew javadoc` fails. Could you please tell me how you build the javadoc?

YizhangCao commented:
I usually generate the Javadoc by running: ./gradlew clean javadoc from the project root directory.

espertusnu commented:
I get a failure when I try running that from commit e8bbe0b174.

ellen@ELLEN_DESKTOP MINGW64 ~/IdeaProjects-espertusnu/jabref ((e8bbe0b174...))
$ ./gradlew clean javadoc

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':jabgui:javadoc'.
> Could not resolve all dependencies for configuration ':jabgui:compileClasspath'.
   > Could not find io.github.adr:e-adr:2.0.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://oss.sonatype.org/content/repositories/snapshots/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://oss.sonatype.org/content/repositories/snapshots/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://jitpack.io/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://jitpack.io/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://oss.sonatype.org/content/groups/public/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://oss.sonatype.org/content/groups/public/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://maven-central.storage-download.googleapis.com/repos/central/data/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://maven-central.storage-download.googleapis.com/repos/central/data/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://sandec.jfrog.io/artifactory/repo/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://sandec.jfrog.io/artifactory/repo/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - file:/C:/Users/ellen/IdeaProjects-espertusnu/jabref/jablib/lib/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - file:/C:/Users/ellen/IdeaProjects-espertusnu/jabref/jablib/lib/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
     Required by:
         project ':jabgui'
   > Could not find io.github.adr:e-adr:2.0.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://oss.sonatype.org/content/repositories/snapshots/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://oss.sonatype.org/content/repositories/snapshots/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://jitpack.io/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://jitpack.io/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://oss.sonatype.org/content/groups/public/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://oss.sonatype.org/content/groups/public/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://maven-central.storage-download.googleapis.com/repos/central/data/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://maven-central.storage-download.googleapis.com/repos/central/data/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - https://sandec.jfrog.io/artifactory/repo/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - https://sandec.jfrog.io/artifactory/repo/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
       - file:/C:/Users/ellen/IdeaProjects-espertusnu/jabref/jablib/lib/io/github/adr/e-adr/2.0.0-SNAPSHOT/maven-metadata.xml
       - file:/C:/Users/ellen/IdeaProjects-espertusnu/jabref/jablib/lib/io/github/adr/e-adr/2.0.0-SNAPSHOT/e-adr-2.0.0-SNAPSHOT.pom
     Required by:
         project ':jabgui' > project :versions

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 12s
9 actionable tasks: 9 up-to-date

YizhangCao commented:
I understood why it has bugs right now, I changed adr:e-adr:2.0.0 to 1.0.0 due to test error but forget to update Javadoc, I will update it as soon as possible. Sorry about my mistake

espertusnu commented:
When you regenerate the javadoc, please tell me the local path to the generated page that has your changes.

YizhangCao committed:
"Fix e-adr dependency issue (downgrade from 2.0.0-SNAPSHOT to 1.0.0)"

YizhangCao committed:
Fix @link problem

YizhangCao commented:
Sorry about that. I mess up. I found I fixed error to branch "contribute-issues676. I have corrected codes and fix error in that but not branch issues676. Right now I regenerate and clean javadoc by run : ./gradlew clean javadoc. It shows: BUILD SUCCESSFUL in 1m 20s
50 actionable tasks: 31 executed, 11 from cache, 8 up-to-date. 
 
The local path to the generated javadoc index page is:
D:\Program\jabref\jablib\build\docs\javadoc\index.html

I also view the javadoc for all modules:
jablib: D:\Program\jabref\jablib\build\docs\javadoc\index.html
jabgui: D:\Program\jabref\jabgui\build\docs\javadoc\index.html
jabkit: D:\Program\jabref\jabkit\build\docs\javadoc\index.html

espertusnu commented:
Thank you! Never be shy about letting me know if you don't understand what I'm asking.

Were you able to find a generated HTML page that corresponded to your class?

YizhangCao committed:
Update IntelliJ inspection profile configuration

YizhangCao commented:
Thanks professor, sometimes I misunderstood what you mean. 

I wasn't able to find generated HTML pages for the classes I modified (AuthorTest.java and AuthorListTest.java).
I realized this is because I modified test classes located in jablib/src/test/java/org/jabref/model/entry/, and the javadoc task only generates documentation for main source files in src/main/java/, not for test files in src/test/java/.
The javadoc that was generated is for the main source code (the production classes), but my changes were refactoring test methods to use @ParameterizedTest, which are in the test source directory.
Should I be looking for something else, or I misunderstand your question again?

espertusnu commented:
No, I think you're correct. Javadoc isn't typically generated for test classes. I was surprised to see javadoc-style comments. It can be confusing because "javadoc" can refer to:
* the comments within source code
* the tool that converts those comments to html
* the html that is generated by the tool

I was thinking of the third definition; you were thinking of the first.

YizhangCao commented:
Thank you for the clarification! That makes sense now. I understand the distinction between the javadoc comments in the source code versus the generated HTML documentation. I appreciate you taking the time to explain the terminology!

espertusnu commented:
I don't think your last commit message matches what you actually committed. Did you mean to commit a change to [‎.idea/inspectionProfiles/Project_Default.xml](https://github.com/YizhangCao/jabref/pull/1/files#diff-bbf0d02ebcf6764cae92fa4045a2d6f15f8a479cd1aa2ab4d1fd435a3aa74bbd)?

YizhangCao commented:
Thank you for catching that! I accidentally committed an IntelliJ IDEA configuration file (.idea/inspectionProfiles/Project_Default.xml) with the wrong commit message. This was from when I was trying to configure the import order settings in IntelliJ. Let me figure it out

espertusnu commented:
That's an easy mistake to make.

Please go to the Files tab and make sure that all changes are intended.

Please avoid force push.

YizhangCao committed:
Remove IDE configuration file

YizhangCao commented:
Thank you! I reviewed the Files tab and found an unintended IDE configuration file (.idea/inspectionProfiles/Project_Default.xml). I've removed it from the PR with a new commit. The PR now only contains the intended parameterized test changes to AuthorTest.java and AuthorListTest.java.

espertusnu commented:
Thanks, but I'm still seeing 11 changed files. Is my view out of date?

<img width="888" height="77" alt="image" src="https://github.com/user-attachments/assets/0e2a5c2b-be5e-403e-ba7a-78e976f4ecce" />

YizhangCao commented:
I see the problem now. My branch has e-adr dependency fixes mixed in with the parameterized test changes. Let me check what files are included and clean this up. Should I create a separate branch with only the AuthorTest.java and AuthorListTest.java changes for issue #676?

espertusnu commented:
Aha!

> Should I create a separate branch with only the AuthorTest.java and AuthorListTest.java changes for issue https://github.com/JabRef/jabref/issues/676?

You'll do that after this one is approved.

Let's try fixing this issue a different way: Put the updates on your main branch too. Since this PR is showing the differences between `main` and `issue676`, that should fix it.

YizhangCao committed:
Trigger PR refresh

YizhangCao commented:
Thanks for suggestions. I have done what exactly you suggest. I've been investigating this issue. Here's what I found:
When I run git diff main..issue676 --name-only locally, it shows only 3 files:

.idea/inspectionProfiles/Project_Default.xml (IDE config file - I'm working on removing this)
AuthorListTest.java (my parameterized test changes)
AuthorTest.java (my parameterized test changes)

However, the PR on GitHub shows 11 files and "31 commits". I added the e-adr dependency fixes to my main branch as you suggested, and pushed successfully. My local git shows everything is in sync with origin/main.
I'm not sure why GitHub is still showing 11 files. The PR says "YizhangCao wants to merge 31 commits into main from issue676" - should it be comparing to my fork's main branch (YizhangCao/jabref:main) or the upstream main (JabRef/jabref:main)?
What should I do to get the PR to show only the 2 test file changes?

espertusnu commented:
Thanks for doing as I requested. I see 11 files changed but just your commits if I go to: https://github.com/YizhangCao/jabref/compare/main...YizhangCao:jabref:issue676, which would compare the latest versions.

espertusnu commented:
Some of your commits have made changes to additional files, such as this one:

https://github.com/YizhangCao/jabref/pull/1/commits/849ba58222fdab33f55a5a3e869582f0314b42fe

Did you mean to check in changes to those files?

I think the way you pulled in the adr fix led to this problem. The next time you need to pull in bug fixes, I can help you through the process. It's tempting to use the quickest fix, but that sometimes leads to later trouble.

espertusnu commented:
My advice to you at this point is to:

Pull the latest version of main from upstream into your main branch and push it to origin.

Start a new branch off main

In that new branch, grab the two changed files from issue 676:

git checkout issue676 path-to-AuthorListTest.java path-to-AuthorTest.java

git status (which should show just those two changed files)

Let me know how it goes

espertusnu commented:
I'm trying to understand this. Could you please not push to main or issue676 for now?

espertusnu commented:
I found out that the way to see on github the same thing we saw from the command line is this:

https://github.com/YizhangCao/jabref/compare/main..issue676

It uses two dots, not 3.

espertusnu commented:
I'm done with my checks. You can make changes to main and any other branches.

YizhangCao commented:
Thank you for the guidance! I've created a clean branch issue676-v2 with only the two test file changes. I do:
Pulled the latest from upstream/main into my main branch
Created a new branch issue676-v2 from main
Cherry-picked only AuthorListTest.java and AuthorTest.java from issue676
Pushed to origin/issue676-final

espertusnu commented:
> Thank you for the guidance! I've created a clean branch issue676-v2 with only the two test file changes. I do: Pulled the latest from upstream/main into my main branch Created a new branch issue676-v2 from main Cherry-picked only AuthorListTest.java and AuthorTest.java from issue676 Pushed to origin/issue676-final

Excellent! It sounds like you're really getting the hang of git.

One nit: "cherry-pick" means something different: It refers to a commit, not a file. I'll talk about it in class on Wednesday.
